artifact.go

```diff
@@ -215,55 +218,413 @@ func (m artifactModule) postBuildArtifactHandler(c *gin.Context) {
 	if !ok {
 		return
 	}
+	files, ok := parseMultipartFormData(c, buildID)
+	if !ok {
+		return
+	}
+	_, ok = m.createArtifacts(c, files, buildID)
+	if !ok {
+		return
+	}
+	c.Status(http.StatusCreated)
+}
+
+type file struct {
+	name     string
+	fileName string
+	data     []byte
+}
+
+// postTestResultDataHandler godoc
+// @summary Post test result data
+// @tags artifact
+// @accept multipart/form-data
+// @param buildid path int true "Build ID"
+// @param file formData file true "Test result artifact file"
+// @success 201 {object} TestResultListSummary "Added new test result data and created summary"
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result-data [post]
+func (m artifactModule) postTestResultDataHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	files, ok := parseMultipartFormData(c, buildID)
+	if !ok {
+		return
+	}
+
+	artifacts, ok := m.createArtifacts(c, files, buildID)
+	if !ok {
+		return
+	}
+
+	summaries := make([]TestResultSummary, 0, len(artifacts))
+	lotsOfDetails := make([]TestResultDetail, 0)
+
+	for _, artifact := range artifacts {
+		summary, details, err := getTestSummaryAndDetails(artifact.Data, artifact.ArtifactID, buildID)
+		if err != nil {
+			log.Warn().
+				WithError(err).
+				WithString("filename", artifact.Name).
+				WithUint("build", buildID).
+				WithUint("artifact", artifact.ArtifactID).
+				Message("Failed to unmarshal; invalid TRX/XML format, or the structure" +
+					" receiving the data is malformed.")
+
+			ginutil.WriteProblemError(c, err,
+				problem.Response{
+					Type:   "/prob/unexpected-response-format",
+					Status: 502,
+					Title:  "Unexpected response format.",
+					Detail: fmt.Sprintf(
+						"Failed parsing test result ID %d, for build with ID %d in"+
+							" database. Expected TRX/XML format. Could also be due to an"+
+							" internal bug.", summary.ArtifactID, buildID),
```

Comment:
This error message projects a lack of confidence :p 

Suggest eluding that "could be due to an internal bug" part. There's always the possibility of a scripting error, but I don't think we should suggest that in the error messages.

Everyone can write bugs, but we should not expect our code to have bugs. If that's the case, then we need to fix that. *(While there are parts of Wharf I expect has a lot of bugs in, but that's nothing we should aim for)*

```suggestion
							" database. Expected TRX/XML format.",
							summary.ArtifactID, buildID),
```